### PR TITLE
Use keyword arguments for performance

### PR DESCRIPTION
**Pull request by bogdan**
Following up for https://github.com/rails/rails/pull/24511#issuecomment-208621285

r? @jeremy

### CODE REVIEW

File: activemodel/lib/active_model/dirty.rb [hunk 1]
```diff
@@ -174,12 +174,10 @@ def changed_attributes
...
     # Handles <tt>*_changed?</tt> for +method_missing+.
-    def attribute_changed?(attr, options = nil) #:nodoc:
+    def attribute_changed?(attr, from: :none, to: :none)
       result = changes_include?(attr)
-      if options
-        result &&= options[:to] == __send__(attr) if options.key?(:to)
-        result &&= options[:from] == changed_attributes[attr] if options.key?(:from)
-      end
+      result &&= to == __send__(attr) unless to == :none
+      result &&= from == changed_attributes[attr] unless from == :none
```
Comment (jeremy) [inline activemodel/lib/active_model/dirty.rb]: Technically `:none` could be a valid to/from value. We'd need a guaranteed unique sentinel value, like `OMITTED = Object.new`. That starts looking pretty bad though.

Our other choice is `**options`, but then we lose the `if options` short-circuit when none are provided.

Comment (bogdan) [inline activemodel/lib/active_model/dirty.rb]: Hard to imagine the case where attribute value is a symbol. Definitely not the case of AR.

**Comment by vipulnsward** [comment]
@bogdan I had sent https://github.com/rails/rails/pull/24516, feel free to pull in the dropping of options change here.

**Comment by bogdan** [comment]
@jeremy used a more exotic symbol to describe the option omission to make impossible to randomly match to real value.

File: activemodel/lib/active_model/dirty.rb [hunk 2]
```diff
@@ -174,12 +174,10 @@ def changed_attributes
     end
 
     # Handles <tt>*_changed?</tt> for +method_missing+.
-    def attribute_changed?(attr, options = nil) #:nodoc:
+    def attribute_changed?(attr, from: :__none__, to: :__none__) # :nodoc:
```
Comment (sgrif) [inline activemodel/lib/active_model/dirty.rb]: Can we use a poison object here instead so that all possible user provided values are valid?

Comment (sgrif) [inline activemodel/lib/active_model/dirty.rb]: e.g. https://github.com/rails/rails/blob/master/activerecord/lib/active_record/attributes.rb#L220

**Comment by bogdan** [comment]
@jeremy @sgrif you want it - you get it.

File: activemodel/lib/active_model/dirty.rb [hunk 3]
```diff
@@ -119,6 +119,9 @@ module Dirty
     extend ActiveSupport::Concern
     include ActiveModel::AttributeMethods
 
+    OPTION_NOT_GIVEN = Object.new
```
Comment (sgrif) [inline activemodel/lib/active_model/dirty.rb]: Should this be `# :nodoc:`?

File: activemodel/lib/active_model/dirty.rb [hunk 4]
```diff
@@ -174,12 +177,10 @@ def changed_attributes
...
 
     # Handles <tt>*_changed?</tt> for +method_missing+.
-    def attribute_changed?(attr, options = nil) #:nodoc:
+    def attribute_changed?(attr, from: OPTION_NOT_GIVEN, to: OPTION_NOT_GIVEN) # :nodoc:
       result = changes_include?(attr)
-      if options
-        result &&= options[:to] == __send__(attr) if options.key?(:to)
-        result &&= options[:from] == changed_attributes[attr] if options.key?(:from)
-      end
+      result &&= to == __send__(attr) unless to == OPTION_NOT_GIVEN
```
Comment (sgrif) [inline activemodel/lib/active_model/dirty.rb]: Do you think this is a good opportunity to break this into multiple lines? It's very difficult to parse mentally in this form.

Comment (bogdan) [inline activemodel/lib/active_model/dirty.rb]: How about this:

``` ruby
changes_include?(attr) && 
  (to == NOT_GIVEN || to == __send__(attr)) &&
  (from == NOT_GIVEN || from == changed_attributes[attr]) 
```

 ?

Comment (sgrif) [inline activemodel/lib/active_model/dirty.rb]: Yes, I think that is much easier to read.

**Comment by sgrif** [comment]
The failures are valid. We'll still need the `!!` here

File: activemodel/lib/active_model/dirty.rb [hunk 5]
```diff
@@ -189,7 +189,7 @@ def attribute_was(attr) # :nodoc:
     end
 
     # Handles <tt>*_previously_changed?</tt> for +method_missing+.
-    def attribute_previously_changed?(attr, options = {}) #:nodoc:
+    def attribute_previously_changed?(attr) #:nodoc:
```
Comment (jeremy) [inline activemodel/lib/active_model/dirty.rb]: What's the story re. changing the method signature? Is it related to this PR or just seem superfluous? Is it used/overridden/expected elsewhere?

File: activemodel/lib/active_model/dirty.rb [hunk 6]
```diff
@@ -174,13 +177,10 @@ def changed_attributes
...
-      if options
-        result &&= options[:to] == __send__(attr) if options.key?(:to)
-        result &&= options[:from] == changed_attributes[attr] if options.key?(:from)
-      end
-      !!result
+    def attribute_changed?(attr, from: OPTION_NOT_GIVEN, to: OPTION_NOT_GIVEN) # :nodoc:
+      changes_include?(attr) &&
+        (to == OPTION_NOT_GIVEN || to == __send__(attr)) &&
+        (from == OPTION_NOT_GIVEN || from == changed_attributes[attr])
     end
```
Comment (jeremy) [inline activemodel/lib/active_model/dirty.rb]: Nice 👍

File: activemodel/lib/active_model/dirty.rb (hunk 5 above)
Comment (bogdan) [inline activemodel/lib/active_model/dirty.rb]: @vipulnsward did it in his alternative PR for some reason and recommended it to be used in mine. Sorry, I didn't think it through myself. @vipulnsward can you explain it?

Comment (vipulnsward) [inline activemodel/lib/active_model/dirty.rb]: @jeremy The options are not passed to, or used here.

Comment (jeremy) [inline activemodel/lib/active_model/dirty.rb]: Looks like the [arg was added for parity with the `attribute_changed?` method signature](https://github.com/rails/rails/pull/19847/files#diff-aaddd42c7feb0834b1b5c66af69814d3R185), despite being unused. It's only been present through 5.0.0 betas. Think we're good to drop it. If the method takes args later, they'll be keyword args anyway.

**Pull request merged**